cirq-core/cirq/experiments/readout_confusion_matrix.py

```diff
@@ -0,0 +1,310 @@
+
+        Args:
+            confusion_matrices: Sequence of Confusion matrices, computed for qubit patterns present
+                                in `measure_qubits`. A single confusion matrix is also accepted.
+            measure_qubits: Sequence of smaller qubit patterns, for which the confusion matrices
+                            were computed. A single qubit pattern is also accepted.
+        Raises:
+            ValueError: If length of `confusion_matrices` and `measure_qubits` is different or if
+                        the shape of any confusion matrix does not match the corresponding qubit
+                        pattern.
+        """
+        if isinstance(confusion_matrices, np.ndarray):
+            confusion_matrices = [confusion_matrices]
+        measure_qubits = cast(
+            Sequence[Sequence['cirq.Qid']],
+            [measure_qubits] if isinstance(measure_qubits[0], ops.Qid) else measure_qubits,
+        )
+        if len(confusion_matrices) != len(measure_qubits):
+            raise ValueError(
+                f"len(confusion_matrices): {len(confusion_matrices)} should be equal to "
+                f"len(measure_qubits): {len(measure_qubits)}"
+            )
+        for i, (cm, q) in enumerate(zip(confusion_matrices, measure_qubits)):
+            if cm.shape != (2 ** len(q),) * 2:
+                raise ValueError(
+                    f"Shape mismatch for confusion matrix {cm} at index {i} corresponding to {q}."
+                    f"Confusion Matrix shape {cm.shape} should match {(2 ** len(q),) * 2}"
+                )
+        self._confusion_matrices = list(confusion_matrices)
+        self._measure_qubits = [list(q) for q in measure_qubits]
+        self._qubits = sorted(set(q for ql in measure_qubits for q in ql))
+
+    @property
+    def confusion_matrices(self) -> List[np.ndarray]:
+        """List of confusion matrices corresponding to `measure_qubits` qubit pattern."""
+        return self._confusion_matrices
+
+    @property
+    def measure_qubits(self) -> List[List['cirq.Qid']]:
+        """Calibrated qubit pattern for which individual confusion matrices were computed."""
+        return self._measure_qubits
+
+    @property
+    def qubits(self) -> List['cirq.Qid']:
+        """Sorted list of all calibrated qubits."""
+        return self._qubits
+
+    def _get_vars(self, qubit_pattern: Optional[Sequence[Sequence['cirq.Qid']]] = None):
+        if qubit_pattern is None:
+            qubit_pattern = self.measure_qubits
+        abcd = "abcdefghijklmnopqrstuvwxyz"
+
+        def qubits_to_abcd(qs: Sequence['cirq.Qid']):
+            assert len(qs) <= len(abcd), "No. of qubits should be <= 26."
+            ret = ''.join(abcd[self.qubits.index(q)] for q in qs)
+            return ret + ret.upper()
+
+        return ','.join(qubits_to_abcd(qs) for qs in qubit_pattern)
+
+    @functools.lru_cache()
+    def _confusion_matrix(self, qubits: Tuple['cirq.Qid']) -> np.ndarray:
```

Comment:
I don't think this is appropriate for lru_cache since `self` isn't hashable, no?

Reply:
self is hashable, as discussed below.